### PR TITLE
Rule S3260: FP when nested private class inherits from outer private class.

### LINKED ISSUE #4983: S3260 does not detect if the class is inherited by a sub-class
### Description

S3260 does not detect if the class is inherited by a sub-class.

### Repro steps

Just paste this example:

```csharp
internal class Test
{
    private class PropertyFetch // S3260 Private classes or records which are not derived in the current assembly should be marked as 'sealed'.
    {
        public virtual bool Fetch() => false;

        private sealed class CustomPropertyFetch : PropertyFetch
        {
            public override bool Fetch() => true;
        }
    }
}
```
And observe that S3260 is displayed while it shouldn't. 
Anyway the class can't be sealed as it has a virtual method.

### Known workarounds

Set the class as abstract (might not be applicable).

### Related information

```xml
<PackageReference Include="SonarAnalyzer.CSharp" Version="8.30.0.37606" />
```

### PR DESCRIPTION
**Pull request by mary-georgiou-sonarsource**
Fixes #4983
Related PR for virtual members FP: #5027

### CODE REVIEW

File: analyzers/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataFactory.cs [hunk 1]
```diff
@@ -41,6 +41,7 @@ public static partial class NuGetMetadataFactory
             "net",
             "netstandard2.1",
             "netstandard2.0",
+            "net6.0",
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/MetadataReferences/NuGetMetadataFactory.cs:44]: wasn't this done by Pavel's PR?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonDerivedPrivateClassesShouldBeSealed.cs [hunk 2]
```diff
@@ -50,16 +50,34 @@ private class PrivateDerivedClassSecondExtension : PrivateDerivedClass { }
 
         private class TheThirdExtension : PrivateDerivedClassSecondExtension { } // Noncompliant
 
-        private class PrivateClass // Noncompliant
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonDerivedPrivateClassesShouldBeSealed.cs:53]: was this previously an FP?

File: analyzers/tests/SonarAnalyzer.UnitTest/Rules/UnusedPrivateMemberTest.cs [hunk 3]
```diff
@@ -249,7 +249,7 @@ public void UnusedPrivateMember_Performance()
         {
             Action verifyAnalyzer = () => Verifier.VerifyAnalyzer(new[] {@"TestCases\UnusedPrivateMember.Performance.cs"},
                                                                   new UnusedPrivateMember(),
-                                                                  GetEntityFrameworkCoreReferences(Constants.NuGetLatestVersion));
+                                                                  GetEntityFrameworkCoreReferences("5.0.12")); // Latest before 6.0.0 for .NET 6 that has Linq versioning collision issue
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/Rules/UnusedPrivateMemberTest.cs:252]: this seems unrelated

**Review by andrei-epure-sonarsource — CHANGES_REQUESTED** [review]
could you please clean the commit history as it seems there are unrelated changes?

otherwise it looks good

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs [hunk 4]
```diff
@@ -26,6 +26,7 @@
 using SonarAnalyzer.Common;
 using SonarAnalyzer.Helpers;
 using StyleCop.Analyzers.Lightup;
+using System.Collections.Generic;
```
Comment (andrei-epure-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs]: I don't understand this commit...

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonDerivedPrivateClassesShouldBeSealed.cs [hunk 5]
```diff
@@ -50,16 +50,34 @@ private class PrivateDerivedClassSecondExtension : PrivateDerivedClass { }
...
+        private class PrivateExternalA { } // Compliant, it's extended in 'ExtendsExternalPrivateClass'
+        private class PrivateExternalB { } // Compliant, it's extended in 'ExtendsExternalPrivateClass'
```
Comment (andrei-epure-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonDerivedPrivateClassesShouldBeSealed.cs:54]:
```suggestion
        private class PrivateExternalB { } // Compliant, it's extended in 'SeriouslyNestedClass'
```

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonDerivedPrivateClassesShouldBeSealed.cs (hunk 2 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/tests/SonarAnalyzer.UnitTest/TestCases/NonDerivedPrivateClassesShouldBeSealed.cs:53]: This particular line no, because this `PrivateClass` was not inherited in any case.

**Comment by mary-georgiou-sonarsource** [comment]
Rebase issue has been fixed.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs (hunk 4 above)
Comment (mary-georgiou-sonarsource) [reply analyzers/src/SonarAnalyzer.CSharp/Rules/NonDerivedPrivateClassesShouldBeSealed.cs]: Rebase issue has been fixed.

**Review by andrei-epure-sonarsource — APPROVED** [review]
LGTM
